### DISCUSSION TITLE
How to get timesheet custom field on invoice template working?

### DISCUSSION THREAD

**Discussion by SaturnusDJ** [kimai/kimai#4007] [Questions & Answers] [answered]
I took the default template, made visual adjustments like column width. That works well.
Getting a custom field on it, does not work.
Extract:
```
    <table class="items">
        <thead>
            <tr>
                <th class="first"><strong>{{ 'date'|trans }}</strong></th>
                <th><strong>{{ 'description'|trans }}</strong></th>
                <!-- <th class="text-right"><strong>{{ 'unit_price'|trans }}</strong></th> -->
                <!-- <th class="text-right"><strong>{{ 'amount'|trans }}</strong></th> -->
                <th class="last text-right"><strong>{{ 'total_rate'|trans }}</strong></th>
            </tr>
        </thead>
        <tbody>
        {% for id, entry in model.calculator.entries %}
            <!-- CONTENT_PART -->
            {% set duration = entry.duration|duration(isDecimal) %}
            {% if entry.fixedRate %}
                {% set rate = entry.fixedRate %}
                {% set duration = entry.amount|amount %}
            {% else %}
                {% set rate = entry.hourlyRate %}
            {% endif %}
            <tr>
                <td nowrap class="first text-nowrap" style="width:15%">{{ entry.begin|date_short }}</td>
                <td contenteditable="true">
                    {% set subproject = entry.getMetaFieldValue('subproject') %}
                    {% if subproject is not empty %}
                        {{ subproject }}
                    {% elseif entry.description is not empty %}
                        {{ entry.description|nl2br }}
                    {% else %}
                        {% if entry.activity is not null %}{{ entry.activity.name }} / {% endif %}{{ entry.project.name }}
                    {% endif %}
                </td>
                <!-- <td nowrap class="text-nowrap text-right">{{ rate|money(currency) }}</td> -->
                <!-- <td nowrap class="text-nowrap text-right">{{ duration }}</td> -->
                <td nowrap class="last text-nowrap text-right" style="width:15%">{{ entry.rate|money(currency) }}</td>
            </tr>
        {% endfor %}
        </tbody>

```

As mentioned on https://www.kimai.org/documentation/invoices.html#twig-templates, the `{% for id, entry in model.calculator.entries %}` part is present. It was already there. Next I basically copy the `timesheet` example and edit it, expanding the if statement from the default invoice template. The custom field is `subproject` under the `Time Tracking` category. What actually happens is that the description is shown on the invoice and if that is empty, the activity. If I force the subproject to be shown, the invoice field is empty. Tried quite some code variants but no custom field is showing up.

Setup in Kimai, as per documentation:
![image](https://user-images.githubusercontent.com/14320018/236227169-37c78420-b091-4be1-bfee-82d1d6096966.png)

What could I be doing wrong here?

Documentation:
>Custom fields
>
>Iterating above all entries (line items) in the invoice with {% for id, entry in model.calculator.entries %} allows access to your custom fields.
>
>Want to use a timesheet custom field in your template?
>
>{% set meta = entry.getMetaFieldValue('foo') %}
>{% if meta is not null %}
>    Foo: {{ meta }}
>{% endif %}
>
>Please be aware:
>
>   - you can access timesheet custom fields only if you use the “sum calculation” standard, which creates one invoice line item per timesheet.
>   - if you group timesheets e.g. by project, you loose access to their custom fields.
>   - entries could be of type “expense” or other types (depending on your used plugins), you can test that with {% if entry.type == 'timesheet' '%}

**Comment by kevinpapst** [comment, marked as answer]
I might have to update the documentation, to stop recommending PHP methods...

The loop should actually be achieved by using the `entries` array:
```
        {% for row in entries %}
                <tr class="record">
                    <td class="text-left text-nowrap">
                        {{ row['entry.begin']|date_short }}
                    </td>
                    <td>
                        Subproject:
                        {% if row['entry.meta.subproject'] is defined and row['entry.meta.subproject'] != '' %}
                           {{ row['entry.meta.subproject'] }}
                        {% else %}
                            ! Missing !
                        {% endif %}
                    </td>
                    <td class="text-right text-nowrap">
                        {{ row['entry.rate'] }}
                    </td>
                    <td class="text-right text-nowrap">
                        {{ row['entry.total'] }}
                    </td>
                </tr>
            {% endif %}
        {% endfor %}
```

Please copy & paste the debug template from the linked documentation page, use it to create the preview invoice.
Then you find all available values and their notation.

**Comment by SaturnusDJ** [comment]
Thanks, got it working now. :)

I did not mention it, but I did use the debug template. Checked again now and found the custom field. Not sure why I did not find them yesterday. I would not have gotten it working without your help though. Appreciated.

**Comment by kevinpapst** [comment]
Discussion do not need to be closed, because then they are then hidden and the next user doesn't find the solution
